### PR TITLE
tsp, support client package

### PR DESCRIPTION
**Pull request by weidongxu-microsoft**
fix https://github.com/Azure/autorest.java/issues/2886

ts code https://github.com/Azure/autorest.java/pull/2889/commits/5a7bb84831d91d978be32f7fdcc94679a5b8a13c
java code https://github.com/Azure/autorest.java/pull/2889/commits/d00b2e26e9c66c074700a2fd2355824e1f24a232

doc update https://github.com/Azure/typespec-azure/pull/1285

`module-info` would export the package as well.
```java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
// Code generated by Microsoft (R) TypeSpec Code Generator.

module com.cadl.server {
    requires transitive com.azure.core;
    exports com.cadl.server;
    exports com.cadl.server.contoso.sub;
}
```

### CODE REVIEW

File: javagen/src/main/java/com/azure/autorest/model/clientmodel/ServiceClient.java [hunk 1]
```diff
@@ -180,10 +183,19 @@ public final String getDefaultCredentialScopes() {
         return defaultCredentialScopes;
     }
 
-    public final boolean builderDisabled() {
+    public final boolean isBuilderDisabled() {
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/model/clientmodel/ServiceClient.java:186]: Just a rename.

File: javagen/src/main/java/com/azure/autorest/model/clientmodel/ServiceClient.java [hunk 2]
```diff
@@ -180,10 +183,19 @@ public final String getDefaultCredentialScopes() {
...
     }
 
+    /**
+     * Gets the package name for builder and wrapper classes.
+     *
+     * @return the package name for builder and wrapper classes.
+     */
+    public String getBuilderPackageName() {
+        return builderPackageName;
+    }
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/model/clientmodel/ServiceClient.java:197]: I call it `builderPackageName` but actually it is used for both builder and wrapper client.

Welcome recommendation.

File: javagen/src/main/java/com/azure/autorest/model/javamodel/JavaPackage.java [hunk 3]
```diff
@@ -102,9 +102,7 @@ public final void addAsyncServiceClient(String packageKeyWord, AsyncSyncClient a
     }
 
     public final void addSyncServiceClient(String packageKeyWord, AsyncSyncClient syncClient) {
-        JavaFile javaFile = javaFileFactory.createSourceFile(packageKeyWord, syncClient.getClassName());
-        Templates.getServiceSyncClientTemplate().write(syncClient, javaFile);
-        addJavaFile(javaFile);
+        addSyncServiceClient(packageKeyWord, syncClient, false);
     }
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/model/javamodel/JavaPackage.java:106]: Only a minor refactor. Logic should be same.

File: typespec-extension/src/code-model-builder.ts [hunk 4]
```diff
@@ -531,9 +531,29 @@ export class CodeModelBuilder {
...
+      let clientName = client.name;
+      let javaNamespace = getJavaNamespace(this.namespace);
+      const clientFullName = client.name;
+      const clientNameSegments = clientFullName.split(".");
+      if (clientNameSegments.length > 1) {
+        clientName = clientNameSegments.at(-1)!;
+        javaNamespace = getJavaNamespace(
+          this.namespace + "." + clientNameSegments.slice(0, -1).join(".").toLowerCase(),
+        );
+      }
```
Comment (weidongxu-microsoft) [inline typespec-extension/src/code-model-builder.ts:543]: break up the name e.g. `Contoso.Sub.ContosoClient`, so client name is `ContosoClient`, and `Contoso.Sub` get appended to default java namespace.

Comment (weidongxu-microsoft) [reply typespec-extension/src/code-model-builder.ts:543]: Not sure how TCGC handles this. My guess is that it is same in the name.

File: typespec-tests/src/main/java/com/cadl/server/contoso/sub/ContosoClient.java [hunk 5]
```diff
@@ -2,7 +2,7 @@
 // Licensed under the MIT License.
 // Code generated by Microsoft (R) TypeSpec Code Generator.
 
-package com.cadl.server;
+package com.cadl.server.contoso.sub;
```
Comment (weidongxu-microsoft) [inline typespec-tests/src/main/java/com/cadl/server/contoso/sub/ContosoClient.java:5]: Result of the test.

For `name=Contoso.Sub.ContosoClient` in below
```ts
  @client({
    name: "Contoso.Sub.ContosoClient",
    service: Cadl.ContosoServer,
  })
```

The Client/AsyncClient/Builder is moved to `contoso.sub` package.

File: typespec-tests/src/main/java/com/encode/bytes/models/Base64urlArrayBytesProperty.java [hunk 6]
```diff
@@ -24,7 +24,7 @@ public final class Base64urlArrayBytesProperty implements JsonSerializable<Base6
      * The value property.
      */
     @Generated
-    private List<Base64Url> value;
+    private final List<Base64Url> value;
```
Comment (haolingdong-msft) [inline typespec-tests/src/main/java/com/encode/bytes/models/Base64urlArrayBytesProperty.java:27]: Just curious, why this becomes `final`, do we do some fix?

**Review by haolingdong-msft — APPROVED** [review]
LGTM

File: typespec-tests/src/main/java/com/encode/bytes/models/Base64urlArrayBytesProperty.java (hunk 6 above)
Comment (weidongxu-microsoft) [reply typespec-tests/src/main/java/com/encode/bytes/models/Base64urlArrayBytesProperty.java:27]: This is due to the removal of customization code https://github.com/Azure/autorest.java/blob/abbac504dfb7894d2ed25d13eb46b39b1b98d891/typespec-tests/customization/src/main/java/CustomizationEncodeBytes.java

Customization is no longer needed for this encode/bypes/models test, after switch to azure-json (Alan probably just kept the customization code).

Comment (haolingdong-msft) [reply typespec-tests/src/main/java/com/encode/bytes/models/Base64urlArrayBytesProperty.java:27]: I see.

**Pull request merged**